ext/nnstreamer/tensor_filter/tensor_filter_caffe2_core.cc

```diff
@@ -0,0 +1,465 @@
+/**
+ * Copyright (C) 2019 Samsung Electronics Co., Ltd. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ */
+/**
+ * @file   tensor_filter_caffe2_core.cc
+ * @author HyoungJoo Ahn <hello.ahn@samsung.com>
+ * @date   31/5/2019
+ * @brief  connection with caffe2 libraries.
+ *
+ * @bug     No known bugs.
+ */
+
+#include <unistd.h>
+#include <algorithm>
+
+#include <nnstreamer_plugin_api.h>
+#include "tensor_filter_caffe2_core.h"
+
+/**
+ * @brief Macro for debug mode.
+ */
+#ifndef DBG
+#define DBG FALSE
+#endif
+
+std::map <char*, Tensor*> Caffe2Core::inputTensorMap;
+
+/**
+ * @brief	Caffe2Core creator
+ * @param	_model_path	: the logical path to '{model_name}.tffile' file
+ * @note	the model of _model_path will be loaded simultaneously
+ * @return	Nothing
+ */
+Caffe2Core::Caffe2Core (const char * _model_path, const char *_model_path_sub)
+{
+  pred_model_path = _model_path;
+  init_model_path = _model_path_sub;
+
+  gst_tensors_info_init (&inputTensorMeta);
+  gst_tensors_info_init (&outputTensorMeta);
+}
+
+/**
+ * @brief	Caffe2Core Destructor
+ * @return	Nothing
+ */
+Caffe2Core::~Caffe2Core ()
+{
+  gst_tensors_info_free (&inputTensorMeta);
+  gst_tensors_info_free (&outputTensorMeta);
+}
+
+/**
+ * @brief	initialize the object with caffe2 model
+ * @return 0 if OK. non-zero if error.
+ *        -1 if the model is not loaded.
+ *        -2 if the initialization of input tensor is failed.
+ *        -3 if the initialization of output tensor is failed.
+ */
+int
+Caffe2Core::init (const GstTensorFilterProperties * prop)
+{
+  if (loadModels ()) {
+    g_critical ("Failed to load model\n");
+    return -1;
+  }
+  if (setInputTensorProp (prop)) {
+    g_critical ("Failed to set input tensor\n");
+    return -2;
+  }
+  if (initInputTensor ()) {
+    g_critical ("Failed to initialize input tensor\n");
+    return -3;
+  }
+  if (setOutputTensorProp (prop)) {
+    g_critical ("Failed to set output tensor\n");
+    return -4;
+  }
+  return 0;
+}
+
+#define initializeTensor(type)\
+do {\
+  ReinitializeTensor (\
+      inputTensor,\
+      {\
+        inputTensorMeta.info[i].dimension[0],\
+        inputTensorMeta.info[i].dimension[1],\
+        inputTensorMeta.info[i].dimension[2],\
+        inputTensorMeta.info[i].dimension[3]\
+      },\
+      at::dtype<type> ().device (CPU)\
```

Comment:
can we infer the device from the model file to be loaded rather than using fixed `CPU`?
Also, what is the default device (if we donot specify any device)?

Reply:
currently, the default device for every nnfw of nnstreamer is CPU,  because there is no package for gpu, include tf, tf-lite. When the gpu packages are updated, we will make another option for it or respective fw, like tf-gpu, caffe2-gpu.

Reply:
ok.